### PR TITLE
Ensure that the test state transitions to complete before the test is considered complete.

### PR DESCRIPTION
**Pull request by poletti-marco**
In some environments the state change events and the "test complete" event are sent over a stream that does not guarantee ordering (e.g. as independent HTTP requests when using package:sse).
Without this change, such tests will fail intermittently, when the "test complete" event happens to be delivered before the state change event.

### CODE REVIEW

File: pkgs/test_api/lib/src/backend/live_test_controller.dart [hunk 1]
```diff
@@ -186,9 +193,15 @@ class LiveTestController {
     return liveTest.onComplete;
   }
 
+  /// Returns a future that completes when the test is complete.
+  ///
+  /// We also wait for the state to transition to Status.complete.
+  Future get onComplete =>
```
Comment (jakemac53) [inline pkgs/test_api/lib/src/backend/live_test_controller.dart]: nit: `Future<void>` (I realize the rest of the file doesn't use this style, but `Future<void>` didn't exist when it was written)

**Comment by jakemac53** [comment]
It looks like the travis failure is just due to needing a `dartfmt`

File: pkgs/test_api/lib/src/backend/live_test_controller.dart (hunk 1 above)
Comment (poletti-marco) [reply pkgs/test_api/lib/src/backend/live_test_controller.dart]: Now added `<void>` in a bunch of places.

**Comment by poletti-marco** [comment]
The tests on Travis CI seem to be failing: https://travis-ci.org/dart-lang/test/builds/527418451
The issues don't seem related to my change (or are they?).
How can I reproduce locally?

I tried setting Travis-like env vars and running "./tool/travis.sh command_0" but I get compiler crashes.
Maybe it's because I've done the internal Dart setup instead of the open source one?

**Comment by grouma** [comment]
The failures are almost certainly related to your change.

It's difficult to emulate the Travis environment entirely. That being said you can easily run the tests locally through `pub run test`.

You can also dig into the travis runs to see the specific changes and just run a single one, e.g. `pub run test test/runner/signal_test.dart`.

I ran a couple of the problematic tests locally and I'm getting the same failures. It looks like the tests are not completing so we must be overlooking some edge case.

**Comment by poletti-marco** [comment]
I think I've found the issue now.
However I get linter errors in Travis CI unrelated with my change.
I also tried to fix those in this PR, but they are still reported.
Could you please advise what to do?

**Comment by poletti-marco** [comment]
Actually it turns out that fixing them in this commit works.

File: pkgs/test/lib/src/runner/node/socket_channel.dart [hunk 2]
```diff
@@ -3,7 +3,7 @@
 // BSD-style license that can be found in the LICENSE file.
 
 @JS()
-library _;
+library library_name;
```
Comment (grouma) [inline pkgs/test/lib/src/runner/node/socket_channel.dart]: Why is this change necessary?

File: pkgs/test_api/lib/src/backend/live_test_controller.dart [hunk 3]
```diff
@@ -186,9 +193,15 @@ class LiveTestController {
     return liveTest.onComplete;
   }
 
+  /// Returns a future that completes when the test is complete.
+  ///
+  /// We also wait for the state to transition to Status.complete.
+  Future<void> get onComplete =>
```
Comment (grouma) [inline pkgs/test_api/lib/src/backend/live_test_controller.dart]: These changes should be outlined in the `CHANGELOG` and we should bump the `pubspec` version.

File: pkgs/test_api/lib/src/frontend/expect.dart [hunk 4]
```diff
@@ -138,8 +138,9 @@ Future _expect(actual, matcher,
 
   var matchState = {};
   try {
-    if ((matcher as Matcher).matches(actual, matchState))
+    if ((matcher as Matcher).matches(actual, matchState)) {
```
Comment (grouma) [inline pkgs/test_api/lib/src/frontend/expect.dart]: Did we add a lint for this recently? Not opposed to the clean up but it makes the PR hard to follow. Feel free to open an issue and I'll just fix all packages in a single go. I'd drop these type of changes from this PR.

Comment (poletti-marco) [reply pkgs/test_api/lib/src/frontend/expect.dart]: It's true that it's unrelated.
Reverted these changes from here, and filed http://go/gh/dart-lang/test/issues/1032.
Once that is fixed I'll merge and make sure that Travis succeeds on this PR.

Comment (grouma) [reply pkgs/test_api/lib/src/frontend/expect.dart]: I see. I didn't realize we now make use of `package:pedantic` for lints. Changes to that package will break us. Fix is here: https://github.com/dart-lang/test/pull/1033

Sorry about that.

File: pkgs/test/lib/src/runner/node/socket_channel.dart (hunk 2 above)
Comment (poletti-marco) [reply pkgs/test/lib/src/runner/node/socket_channel.dart]: To make a lint check happy. Now reverted together with the rest.

File: pkgs/test_api/lib/src/frontend/expect.dart (hunk 4 above)
Comment (poletti-marco) [reply pkgs/test_api/lib/src/frontend/expect.dart]: The fix is in (thanks!), merged now.

File: pkgs/test_api/lib/src/backend/live_test_controller.dart (hunk 3 above)
Comment (poletti-marco) [reply pkgs/test_api/lib/src/backend/live_test_controller.dart]: Done.

**Comment by poletti-marco** [comment]
Travis passed, this is ready to merge.
